iModelCore/ECDb/Tests/NonPublished/ClassViewsTests.cpp

```diff
@@ -0,0 +1,1551 @@
+                ASSERT_EQ(ECSqlStatus::Success, v["d"].BindDouble(d_array[n]));
+                ASSERT_EQ(ECSqlStatus::Success, v["dt"].BindDateTime(dt_array[n]));
+                ASSERT_EQ(ECSqlStatus::Success, v["dtUtc"].BindDateTime(dtUtc_array[n]));
+                ASSERT_EQ(ECSqlStatus::Success, v["i"].BindInt(i_array[n]));
+                ASSERT_EQ(ECSqlStatus::Success, v["l"].BindInt64(l_array[n]));
+                ASSERT_EQ(ECSqlStatus::Success, v["s"].BindText(s_array[n].c_str(), IECSqlBinder::MakeCopy::No));
+                ASSERT_EQ(ECSqlStatus::Success, v["p2d"].BindPoint2d(p2d_array[n]));
+                ASSERT_EQ(ECSqlStatus::Success, v["p3d"].BindPoint3d(p3d_array[n]));
+                ASSERT_EQ(ECSqlStatus::Success, v["geom"].BindGeometry(*geom_array[n]));
+            }
+        }
+        if (auto o = &stmt.GetBinder(++idx)) {
+            for (int n = 0; n < 2;++n) {
+                auto& v = o->AddArrayElement();
+                for(auto& m : b_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["b_array"].AddArrayElement().BindBoolean(m));
+                for(auto& m : bi_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["bi_array"].AddArrayElement().BindBlob((void const*)&m[0], (int)m.size(), IECSqlBinder::MakeCopy::No));
+                for(auto& m : d_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["d_array"].AddArrayElement().BindDouble(m));
+                for(auto& m : dt_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["dt_array"].AddArrayElement().BindDateTime(m));
+                for(auto& m : dtUtc_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["dtUtc_array"].AddArrayElement().BindDateTime(m));
+                for(auto& m : i_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["i_array"].AddArrayElement().BindInt(m));
+                for(auto& m : l_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["l_array"].AddArrayElement().BindInt64(m));
+                for(auto& m : s_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["s_array"].AddArrayElement().BindText(m.c_str(), IECSqlBinder::MakeCopy::No));
+                for(auto& m : p2d_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["p2d_array"].AddArrayElement().BindPoint2d(m));
+                for(auto& m : p3d_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["p3d_array"].AddArrayElement().BindPoint3d(m));
+                for(auto& m : geom_array)
+                    ASSERT_EQ(ECSqlStatus::Success, v["geom_array"].AddArrayElement().BindGeometry(*m));
+            }
+        }
+        ASSERT_EQ(BE_SQLITE_DONE, stmt.Step(instKey));
+        m_ecdb.SaveChanges();
+    }
+   //! HARD_CODED_IDS
+   // do not indent the following string as it increases the size of string and compiler will error out.
+    BeJsDocument expected;
+    expected.Parse(R"json([
+  [
+    "0x1",
```

Comment:
Should this expected json be stored in its own file?

Reply:
Whenever possible test should be fully atomic with no external dependency.  (_This is something I & Krischan found out when we wasted a lot of time on fixing the test and realized the test should be atomic and self-contained with little as possible external dependency_). This strategy prevents the test from breaking for external reasons.  It also makes it easy to update, move, and debug tests. Some take away from our experience was
1. Data for the test should be part of the test. In our case Xml schemas and output data.
2. Assert should be in the test body. If you have an array of data you want to assert use a single assert for each case. Instead of using a loop to assert. This is to avoid figuring out which loop iteration is causing assert failure.
3. Do not assert an external re-usable function. This makes it harder to figure out where the test is failing.
4. The test should be flat with no function calls to reusable test functions with asserts or loops or external data loaded from files.
5. If the test is self-contained and when it's deleted it will have no dependency floating around that needs to be deleted as well.

Pasting those large json in the test might make it less readable but it still makes the test have zero dependencies.  Imagine when I change code in ECDb and this test fails and all you see is a test.json file getting updated in PR with no context which test it fixed or was failing. Since PR will not connect the data file to the test.

Reply:
closing comment not changing this json for now